### DISCUSSION TITLE
Read values from relationship chain

### DISCUSSION THREAD

**Discussion by gerhardmpl** [Budibase/budibase#5191] [Help] [answered]
I have three Budibase tables linked with many-to-many relationships and I need to work with values along the relationship chain on a screen. 

1. The ```information``` table holds information (like _lead_, _invoice_ or _order_) and every information is classified in terms of confidentiality (_pubilc_, _internal_, _confidential_, - shown as numbered circles in the diagram).
2. The ```service``` tables holds services (like the _CRM Service_) that process the information (like _lead_) and inherits the maximum confidentiality of related information.
3. The ```asset```tabels holds the assets (like _server_, _storage_) that are needed to run the service and inherits the maximum confidentiality of related services (which results from the information processed).

I hope the app export and diagram helps to understand the dependencies (_relationship chain_) and miximal value principle.

[relationship-export-1648727355640.txt](https://github.com/Budibase/budibase/files/8389010/relationship-export-1648727355640.txt)
![grafik](https://user-images.githubusercontent.com/40722408/161047636-a2eafee4-0f4b-4b94-9bfd-f50161245d3a.png)

When I read or update an asset, I want to know the current inherited confidentiality and therefore must go through the relationship to read the maximum confidentiality from the information level. As an example, the asset ```SAN``` inherits a value of ```4``` from _customer rating_ through the _CRM Service_. If I change _customer rating_ from ```4``` to ```1```, the asset ```SAN``` inherits a value of ```3``` from _order_ through the _ERP Service_.

The app is attached and in the ```Change Service``` screen you can see how I calculate the inherited maximum confidentiality from the related information with javascript. I think this should also be possible through the relationship chain with some nested data provider and repeater in the ```Change Asset``` screen, but I am lacking the budibase-fuu to achieve that.

Could somebody help me with the components and code or point me in the right direction?

**Comment by melohagan** [comment, marked as answer]
Hi @gerhardmpl 

Thank you for providing a nice diagram and the app export :)

It may be useful to use **formulas** for this: (https://docs.budibase.com/docs/formula)

Assuming the confidentiality can only be directly altered on the information table, then instead of transforming the data in the design layer, you can do it in the data layer. 

So in your *asset* table you could replace the *c_inh* column with the following:

![Screenshot 2022-04-04 at 09 33 18](https://user-images.githubusercontent.com/101575380/161505622-13301fa7-803c-4a8a-831d-31c83aae2894.png)

```
let numberOfServices = $("service").length;
let max = -1;
for (let i = 0; i < numberOfServices; i++) {
	let current = $("service." + i + ".c_inh");
	if (current > max) {
		max = current;
	}
}
return max;
```

Then you would also do the same in the *service* table, but the code would be `$("information")...`

Also when I was doing this, **the formula in the *service* table had to be *static* and the formula in the *asset* table had to be *dynamic***.

**Reply by gerhardmpl** [reply]
@melohagan thank you for taking your time and your great suggestion. Until now, I thought that _formulars_ only work when creating, reading or updating a row through the _design layer_. 

So in my example of changing _customer rating_ from ```4``` to ```1```, I was under the impression that the _CRM service_ row needs to be updated to persist the inherited max _c_inh_ of ```2``` from _lead_ so that the _SAN_ gets updated to ```3```. I hope that makes sense to you because I am a bit jetlagged and probably should first explorer your solution.

Never thought that _formulars_ also work for queries on the data layer.

**Reply by melohagan** [reply]
You're welcome. Hopefully that works out for you, but let me know how it goes either way!

**Reply by gerhardmpl** [reply]
@melohagan While testing your suggestion, I came accross a problem that I can not wrap my head around. If a _service_ has no related _information_ (yet), the above JS script exits with a `Error while executing JS` error.

UPDATE - managed to get around the error with 
```
if ( typeof $("information") === 'undefined' ) {
	return 'is undefined';
} else {
	return 'is definded';
};
```
Sorry I asked too soon.

**Comment by gerhardmpl** [comment]
@melohagan Maybe I didn't do it quite right, but the _asset_ confidentiality seems not to update properly. I configured _information_, _service_ and _asset_ as described in the above diagram. _service_ is ok (`ERP service` inherits a `3`, `CRM service` a `4`), but `SAN` in _asset_ should inherit a `4` (inherited from `CRM service` over `ERP service`), but shows a `3`.

Another problem is, that if you unlink an _information_ from a _service_ (e.g. _lead_ and _customer rating_ from `CRM service`), the _service_ is not updated.

I would apprechiate any help or suggestions on how to read values.
[relationship-export-1649679369408.txt](https://github.com/Budibase/budibase/files/8463982/relationship-export-1649679369408.txt)

**Reply by melohagan** [reply]
Hi @gerhardmpl 

For the asset formula, the max value was being returned inside the for loop instead of outside. Apologies if that was my typo! 

The correct code is:
```
if ( typeof $("service") === 'undefined' ) {
	return 0;
} else {
  let numberOfServices = $("service").length;
  let max = -1;
  for (let i = 0; i < numberOfServices; i++) {
	  let current = $("service." + i + ".c_inh");
	  if (current > max) {
		  max = current;
	  };
  };
  return max;
};
```

**Reply by melohagan** [reply]
Also a possible refactor using null propagation:
```
  let numberOfServices = $("service")?.length;
  let max = 0;
  for (let i = 0; i < numberOfServices; i++) {
	  let current = $("service." + i + ".c_inh");
	  if (current > max) {
		  max = current;
	  };
  };
  return max;
```

**Reply by gerhardmpl** [reply]
Heck, I saw that on _service_ myself but forgot to change it for _asset_. Definitely need more coffee =8-/.
